README.md
grunt-angular-toolbox
=====================

[![Build Status](https://travis-ci.org/Jimdo/grunt-angular-toolbox.svg?branch=master)](https://travis-ci.org/Jimdo/grunt-angular-toolbox)
[![Dependency Status](https://david-dm.org/Jimdo/grunt-angular-toolbox.svg)](https://david-dm.org/Jimdo/grunt-angular-toolbox)

Collection of grunt tasks and opinionated configuration
for development of small and mid-sized angular modules.

Compared to other tooling collections like [angular-seed](https://github.com/angular/angular-seed) and [yeoman angular generator](https://github.com/yeoman/generator-angular), this project aims to produce [bower](http://bower.io/) packages consumable by other angular apps.

Examples:
 - [angular-fontselect](https://github.com/Jimdo/angular-fontselect)
 - [angular-kachelei](https://github.com/Xiphe/angular-kachelei)


[Changelog](https://github.com/Jimdo/grunt-angular-toolbox/blob/master/CHANGELOG.md)
------------------------------------------------------------------------------------

Getting Started
---------------

This plugin requires Grunt `~0.4.5`

If you haven't used [Grunt](http://gruntjs.com/) before, you probably don't want to use this
but build your own tooling.

```sh
npm install grunt grunt-cli --save-dev
npm install grunt-angular-toolbox --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-angular-toolbox');
```

Use
---

```js
// Gruntfile.js
module.exports = function(grunt) {
	'use strict';

	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),
		angularToolbox: { /* see config */ }
		/* project specific configuration here */
	});

	/* load angular-toolbox collection 
		 see Included 3rd party tasks */
	grunt.loadNpmTasks('grunt-angular-toolbox');

	/* custom tasks and hooks */
	grunt.registerTask('default', ['test']);
	grunt.registerTask('build:after', function() {
		grunt.log.ok('work complete!');
	});
};
```


Tasks
-----

Depending on the tasks added by [angularToolbox.tasks](#Config) these grunts are available
in the project.

```sh
#see this list by using
grunt --help
``` 

### $ `grunt test[:(unit|e2e)]`

Run the tests.

__Environment Variables__:  
 - `E2E_SANDBOX_PORT` change port (default: 8765)
 - `KARMA_BROWSERS` overwrite browsers for unit tests (default: PhantomJs,Chrome,Firefox)
 - `KARMA_REPORTERS` overwrite reporters for unit tests (default: progress)
 - `PROTRACTOR_BROWSERS` overwrite browsers for unit tests (default: Chrome)
 - `PROTRACTOR_REPORTERS` overwrite reporters for unit tests (default: dots)

__Options__:  
 - `--browsers` change browsers for current suite(s)
 - `--reporters` change reporters for current suite(s)
 - `--no-coverage` disable coverage reports and instrumentation (useful for debugging)
 - `--no-jshint` disable jshint (useful for debugging)

### $ `grunt demo[:e2e]`
Serve demo or e2e standbox application

__Environment Variables__:  
 - `DEMO_PORT` change port (default: 8000)
 - `E2E_SANDBOX_PORT` change port (default: 8765)

__Options__:  
 - `--port` change port of current task

### $ `grunt coverage`
Serve coverage report, requires `grunt test:unit` to have been run once.

__Environment Variables__:  
 - `COVERAGE_PORT` change port (default: 7000)

__Options__:  
 - `--port` change port

### $ `grunt build[:watch]`
Concatenate, annotate and minify JavaScript and less/sass files
Optionally watch the `src` files and rebuild on change

### $ `grunt release`
Run tests, (if successful) bump version build project, commit changes and push to origin


Default project structure
-------------------------

```
 ┌ demo/
 │ └ index.html
 ├ dist/
 ├ src/
 │ ├ js/
 │ │ ├ *.js (project related js files)
 │ │ └ helper.module.js (initiator of angular module)
 │ ├ less/
 │ │ └ *.less (project related less files)
 │ ├ partials/
 │ │ └ *.html (views for directives)
 │ └ sass/
 │   └ *.scss (project related sass files)
 ├ test/
 │ ├ e2e/
 │ │ ├ env/
 │ │ │ └ index.html
 │ │ ├ SpecHelper.js|coffee (test setup stuff)
 │ │ └ *Spec.js|coffee (end to end test files)
 │ └ unit/
 │   └ *.js|coffee (test setup stuff)
 ├ .jshintrc (optional)
 ├ bower.json (optional)
 ├ package.json
 └ Gruntfile.js|coffee
```

This is customizable via the [config](#config)


Config
------

Values are defaults or explanations. 

```js
// Gruntfile.js
module.exports = function(grunt) {
	'use strict';

	grunt.initConfig({
		'angular-toolbox': {
				/* specify the preconfigured tasks that 
				 should be used in the project */
			tasks: [
				'build',
				'coverage',
				'demo',
				'demo:e2e',
				'release',
				'test'
			],
			
			// whether or not the author in package.json should be set to the
			// contributor with the most commits
			dynamicAuthor: false,
			
			// customize the demo/test environment files
			envFilter: function(env) { return env; },

			// customize project structure
			files: {
				bower: 'bower.json',
				pkg: 'package.json',
				src: {
					js: [
						'src/js/**/helper*.js',
						'src/js/**/!(helper)*.js'
					],
					less: [
						'src/less/**/!(_)*.less'
					],
					sass: [
						'src/sass/**/!(_)*.scss'
					]
				},
				test: {
					unit: [
						'test/unit/**/*.+(js|coffee)'
					],
					e2e: [
						'test/e2e/SpecHelper.+(js|coffee)',
						'test/e2e/**/*Spec.+(js|coffee)'
					]
				},
				styleCheck: [] // run through jshint and jscs
			},

			folder: {
				dist: 'dist/',
				demo: 'demo/',
  			partials: 'src/partials/',
  			e2eSandbox: 'test/e2e/env/',
  			src: {
	  			js: 'src/js/',
	  			sass: 'src/sass/',
	  			less: 'src/less/',
  			}
			},


			// how much commits make a maintainer?
			maintainersThreshold: 15,

			// custom middleware for e2e and demo
			// can be function or array of functions
			customMiddleware: false,

			// the angular module name in case it differs from project name 
			moduleName: 'camelCased "name" from package.json',

			// banners an wraps for generated dist files
			// see https://github.com/Jimdo/grunt-angular-toolbox/tree/master/lib/templates
			templates: {
				banner: '/* ... */',
				bannerMin: '/* ... */',
				wrapTop: '/* ... */',
				wrapBottom: '/* ... *'
			},

			// disable coverage reports
			coverage: true,

			// disable jshint
  		jshint: true,

  		// default port for demo
  		demoPort: 8000,

  		// default port for coverage
  		coveragePort: 7000,

  		// default port for e2e
  		e2eSandboxPort: 8765,

  		// default browsers for karma
  		unitBrowsers: ['Chrome', 'Firefox', 'PhantomJS'],

  		// default browsers for protractor
  		e2eBrowsers: ['Chrome'],

  		// default reporters for karma
  		unitReporters: ['progress'],

  		// default reporter for protractor
  		e2eReporter: 'spec',
		}

		/* additional configuration ... */
	});

	/* rest of gruntfile ... */
};
```


Custom tasks and hooks
----------------------

Register custom tasks and or setup before the added tasks run.

```js
// Gruntfile.js
module.exports = function(grunt) {
	'use strict';

	grunt.initConfig({ /* ... */ });

	/* initiation of tasks ... */

	/* add any custom tasks */
	grunt.registerTask('sayYolo', function() {
		console.log('YOLO!');
	});

	/* hook it into tooling tasks ones.
		 this will be called before all other release tasks */
	grunt.registerTask('release:before', ['sayYolo']);
};
```

### Hooks

 - test:before
 - test:unit:before
 - test:e2e:before
 - test:after
 - test:unit:after
 - test:e2e:after
 - runtest:before
 - runtest:unit:before
 - runtest:e2e:before
 - runtest:after
 - runtest:unit:after
 - runtest:e2e:after
 - demo:before
 - demo:e2e:before
 - coverage:before
 - build:before
 - build:after
 - release:before
 - release:after


Included 3rd party tasks
------------------------

 - [grunt-angular-templates](https://github.com/ericclemmons/grunt-angular-templates)
 - [grunt-bump](https://github.com/vojtajina/grunt-bump)
 - [grunt-concurrent](https://github.com/sindresorhus/grunt-concurrent)
 - [grunt-contrib-concat](https://github.com/gruntjs/grunt-contrib-concat)
 - [grunt-contrib-connect](https://github.com/gruntjs/grunt-contrib-connect)
 - [grunt-contrib-cssmin](https://github.com/gruntjs/grunt-contrib-cssmin)
 - [grunt-contrib-jshint](https://github.com/gruntjs/grunt-contrib-jshint)
 - [grunt-contrib-less](https://github.com/gruntjs/grunt-contrib-less)
 - [grunt-contrib-uglify](https://github.com/gruntjs/grunt-contrib-uglify)
 - [grunt-contrib-watch](https://github.com/gruntjs/grunt-contrib-watch)
 - [grunt-injector](https://github.com/klei/grunt-injector)
 - [grunt-istanbul](https://github.com/taichi/grunt-istanbul)
 - [grunt-karma](https://github.com/karma-runner/grunt-karma)
 - [grunt-ng-annotate](https://github.com/mzgol/grunt-ng-annotate)
 - [grunt-protractor-coverage](https://github.com/r3b/grunt-protractor-coverage)
 - [grunt-protractor-webdriver](https://github.com/seckardt/grunt-protractor-webdriver)
 - [grunt-sass](https://github.com/sindresorhus/grunt-sass)
 - [grunt-shell](https://github.com/sindresorhus/grunt-shell)
 - [grunt-update-contributors](https://github.com/Xiphe/grunt-update-contributors)
 - [grunt-wiredep](https://github.com/stephenplusplus/grunt-wiredep)


LICENSE
-------

> The MIT License
> 
> Copyright (c) 2014 Jimdo GmbH http://jimdo.com
> 
> Permission is hereby granted, free of charge, to any person obtaining a copy
> of this software and associated documentation files (the "Software"), to deal
> in the Software without restriction, including without limitation the rights
> to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
> copies of the Software, and to permit persons to whom the Software is
> furnished to do so, subject to the following conditions:
> 
> The above copyright notice and this permission notice shall be included in
> all copies or substantial portions of the Software.
> 
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
> IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
> FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
> AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
> LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
> OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
> THE SOFTWARE.


CHANGELOG.md
Changelog
=========


1.0.0
-----

Rework everything in order to use [grunt-thrall](https://github.com/Xiphe/Thrall)
Feature-Set is mostly the same.

### Breaking Changes

 - ### removed included bower components
  In order to NOT tie actual dependencies to the tooling,
  toolbox does not include angular bower components anymore.

  __Migration:__
  - add a `bower.json` to the project root.
  - call `bower install angular --save && bower install angular-mocks jasmine-moar-matchers --save-dev`

 - ### `tdd` task removed
  In order to keep the test definition D.R.Y.

  __Migration:__
  - use `grunt test --watch` instead

 - ### configuration api has been changed massively
  - key is now `angularToolbox` instead of `angular-toolbox`
  - see [config](https://github.com/Jimdo/grunt-angular-toolbox#config) for details

 - ## grunt coveralls removed
  in order to not be fixed on one specific "coverage badge provider"

  __Migration:__

  `npm install grunt-coveralls --save-dev`

  ```js
  // Gruntfile.js
  grunt.initConfig({
   /* ... */
   coveralls: {
     src: '<%= angularToolbox.folder.coverageReport %>/lcov.info')
   }
  });
  ```

 - ### jshintrc option removed
  in order to encourage use of `.jshintrc` files

  __Migration:__

  - use `.jshintrc` files

 - ### envFilter removed
  need a new concept here

 - ### default moduleName value (from package.json) will now be camelCased
  in order to need less configuration

  package `grunt-angular-toolbox-sample` will now use `gruntAngularToolboxSample` 
  as module for ngTemplates

  __Migration:__

  when you previously didn't had to set moduleName in config,
  you probably need to do so now.

  ```js
  // Gruntfile.js
  grunt.initConfig({
    /* ... */
    angularToolbox: {
      moduleName: 'my-module'
    }
  });
  ```
 - ### same middleware for demo and e2e-env
  in order to have the same behaviour for both environments

  __Migration:__

  - none

 - ### bower packages will be added automatically
  using [grunt-wiredep](https://github.com/stephenplusplus/grunt-wiredep)
  the vendor configuration has been removed.

  __Migration__

  add this into the `<head>` of demo and e2e sandbox `index.html`

  ```html
  <!-- bower:css -->
  <!-- endbower -->
  <!-- injector:scss -->
  <!-- endinjector -->
  <!-- injector:less -->
  <!-- endinjector -->
  ```

  and this before the `<body>` closes.

  ```html
  <!-- bower:js -->
  <!-- endbower -->
  <!-- injector:js -->
  <!-- endinjector -->
  ```


lib/index.js
// var helpers = require('./helpers');
var grunt = require('grunt');
var helpers = require('./helpers');
var availableTasks = {};

grunt.registerTask('', 'this is a place holder for undefined tasks *', []);

/**
 * Hooks that may be overwritten in order to
 * prepare or tear down additional services.
 */
var hooks = {
  demo: function() {
    grunt.registerTask('demo:before', 'overwrite this *', []);
  },
  'demo:e2e': function() {
    grunt.registerTask('demo:e2e:before', 'overwrite this *', []);
  },
  test: function() {
    grunt.registerTask('test:before', 'overwrite this *', []);
    grunt.registerTask('test:unit:before', 'overwrite this *', []);
    grunt.registerTask('test:unit:after', 'overwrite this *', []);
    grunt.registerTask('test:e2e:before', 'overwrite this *', []);
    grunt.registerTask('test:e2e:after', 'overwrite this *', []);
    grunt.registerTask('test:after', 'overwrite this *', []);
  },
  tdd: function() {
    grunt.registerTask('tdd:before', 'overwrite this *', []);
    grunt.registerTask('tdd:unit:before', 'overwrite this *', []);
    grunt.registerTask('tdd:e2e:before', 'overwrite this *', []);
  },
  coverage: function() {
    grunt.registerTask('coverage:before', 'overwrite this *', []);
  },
  build: function() {
    grunt.registerTask('build:before', 'overwrite this *', []);
    grunt.registerTask('build:after', 'overwrite this *', []);
  },
  release: function() {
    grunt.registerTask('release:before', 'overwrite this *', []);
    grunt.registerTask('release:after', 'overwrite this *', []);
  },
  coveralls: function() {
    grunt.registerTask('coveralls:before', 'overwrite this *', []);
    grunt.registerTask('coveralls:after', 'overwrite this *', []);
  }
};


/**
 * Following tasks can be registered by calling
 * loadTasks({tasks: [ <tasklist> ]})
 */

availableTasks.coveralls = {
  desc: [
    'send coverage report to coveralls'
  ].join('\n'),
  hooks: hooks.coveralls,
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      'coveralls'
    ]));
  }
};

availableTasks.demo = {
  desc: [
    'serve demo application',
    '  options:',
    '    --port  (default: 8000)',
    '  Environment Variables:',
    '    DEMO_PORT  (default: 8000)'
  ].join('\n'),
  hooks: hooks.demo,
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      'shell:bower',
      'ngtemplates:dist',
      'connect:demo',
      'concurrent:demo',
      'watch:partials',
      'watch:demo',
      'shell:openDemo',
      'ngtemplates:dist'
    ]));
    grunt.registerTask('demo', availableTasks.demo.desc, [
      'demo:before',
      'shell:bower',
      'ngtemplates:dist',
      'connect:demo',
      'concurrent:demo'
    ]);
  }
};

availableTasks['demo:e2e'] = {
  desc: [
    'serve e2e sandbox application',
    '  options:',
    '    --port  (default: 8765)',
    '  Environment Variables:',
    '    E2E_SANDBOX_PORT  (default: 8765)'
  ].join('\n'),
  hooks: hooks['demo:e2e'],
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      'shell:deleteInstrumented',
      'instrument',
      'shell:bower',
      'ngtemplates:dist',
      'concurrent:demoE2e',
      'shell:openDemoE2e',
      'connect:test'
    ]));
    grunt.registerTask('demo:e2e', availableTasks['demo:e2e'].desc, [
      'demo:e2e:before',
      'shell:deleteInstrumented',
      'instrument',
      'shell:bower',
      'ngtemplates:dist',
      'concurrent:demoE2e'
    ]);
  }
};

availableTasks.test = {
  desc: [
    'run the tests',
    '  subtasks:',
    '    :unit  execute just unit suite',
    '    :e2e   execute just end to end suite',
    '  options:',
    '    --browsers     set browsers for current suite(s)',
    '    --reporters    set reporters for current suite(s)',
    '    --no-coverage  disable coverage reports and instrumentation (useful for debugging)',
    '    --no-jshint    disable jshint (useful for debugging)',
    '  Environment Variables:',
    '    E2E_SANDBOX_PORT     (default: 8765)',
    '    KARMA_BROWSERS       (default: PhantomJs,Chrome,Firefox)',
    '    KARMA_REPORTERS      (default: progress)',
    '    PROTRACTOR_BROWSERS  (default: Chrome)',
    '    PROTRACTOR_REPORTERS (default: dots)'
  ].join('\n'),
  hooks: hooks.test,
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      helpers.withCoverage('shell:deleteCoverages'),
      'shell:bower',
      helpers.withJshint('jshint'),
      'ngtemplates:dist',
      'karma:all',
      'shell:updateWebdriver',
      helpers.withCoverage('shell:deleteInstrumented'),
      helpers.withCoverage('instrument'),
      helpers.withCoverage('makeReport'),
      'connect:test',
      helpers.withCoverage('protractor_coverage:single', 'protractor:single')
    ]));
    grunt.registerTask(
      'test',
      availableTasks.test.desc,
      function(suite) {
        var tasks = [
          'test:before',
          helpers.withCoverage('shell:deleteCoverages'),
          'shell:bower',
          helpers.withJshint('jshint'),
          'ngtemplates:dist'
        ];

        if (!suite || suite === 'unit') {
          process.env.defaultBrowsers = 'Firefox,Chrome';
          tasks.push(
            'test:unit:before',
            'karma:all',
            helpers.withCoverage('_normalizeUnitCoverage'),
            'test:unit:after'
          );
        }
        if (!suite || suite === 'e2e') {
          tasks.push(
            'test:e2e:before',
            'shell:updateWebdriver',
            helpers.withCoverage('shell:deleteInstrumented'),
            helpers.withCoverage('instrument'),
            'connect:test',
            helpers.withCoverage('protractor_coverage:single', 'protractor:single'),
            'test:e2e:after'
          );
        }

        tasks.push(helpers.withCoverage('makeReport'), 'test:after');
        grunt.task.run(tasks);
      }
    );
  }
};

availableTasks.tdd = {
  desc: availableTasks.test.desc.replace(
    'run the tests',
    'run the tests and rerun on src changes'
  ),
  hooks: function() {
    hooks.test();
    hooks.tdd();
  },
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      helpers.withCoverage('shell:deleteCoverages'),
      'shell:bower',
      'ngtemplates:dist',
      'karma:watch',
      'shell:updateWebdriver',
      'connect:test',
      'protractor_webdriver',
      'shell:triggerTests',

      'watch:andTestBoth',
      'watch:andTestUnit',
      'watch:andTestE2e',

      helpers.withCoverage('shell:deleteCoverages'),
      helpers.withCoverage('shell:deleteInstrumented'),
      helpers.withCoverage('instrument'),
      helpers.withCoverage('protractor_coverage:tdd', 'protractor:tdd'),
      helpers.withCoverage('makeReport')
    ]));
    grunt.registerTask(
      'tdd',
      availableTasks.tdd.desc,
      function(suite) {
        var tasks = ['tdd:before', 'shell:bower'];

        var watcher = '';
        if (!suite || suite === 'unit') {
          tasks.push('tdd:unit:before', 'karma:watch:start');
          watcher = 'watch:andTestUnit';
        }

        if (!suite || suite === 'e2e') {
          tasks.push(
            'tdd:e2e:before',
            'shell:updateWebdriver',
            'connect:test',
            'protractor_webdriver'
          );
          watcher = 'watch:andTestE2e';
        }
        if (!suite) {
          watcher = 'watch:andTestBoth';
        }
        if (!suite || suite === 'unit') {
          tasks.push('shell:triggerTests');
        }

        tasks.push(watcher);
        grunt.task.run(tasks);
      }
    );
  }
};

availableTasks.coverage = {
  desc: [
    'serve coverage report',
    'requires `grunt test` to have been run once',
    '  options:',
    '    --port  (default: 7000)',
    '  Environment Variables:',
    '    COVERAGE_PORT  (default: 7000)'
  ].join('\n'),
  hooks: hooks.coverage,
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      'connect:coverage'
    ]));
    grunt.registerTask('coverage', availableTasks.coverage.desc, [
      'coverage:before',
      'connect:coverage'
    ]);
  }
};

availableTasks.build = {
  desc: [
    'Concatenate, annotate and minify JavaScript and less files',
    '  subtasks:',
    '    :watch rebuild on src change'
  ].join('\n'),
  hooks: hooks.build,
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      'shell:bower',
      'ngtemplates:dist',
      'less:dist',
      'sass:dist',
      'concat:dist',
      'concat:distStyles',
      'ngAnnotate:dist',
      'uglify:dist',
      'concat:distStylesMin',
      'cssmin:dist',
      'watch:build'
    ]));

    grunt.registerTask(
      'build',
      availableTasks.build.desc,
      function(watch) {
        var tasks = [
          'build:before',
          'shell:bower',
          'ngtemplates:dist',
          'less:dist',
          'sass:dist',
          'concat:dist',
          'concat:distStyles',
          'ngAnnotate:dist',
          'uglify:dist',
          'concat:distStylesMin',
          'cssmin:dist',
          'build:after'
        ];

        if (watch === 'watch') {
          tasks.push('watch:build');
        }

        grunt.task.run(tasks);
      }
    );
  }
};

availableTasks.release = {
  desc: [
    'Run tests,',
    '(if successful) bump version build project,',
    'commit changes and push to origin.',
    '  options:',
    '    <see test>',
    '  Environment Variables:',
    '    <see test>'
  ].join('\n'),
  hooks: function() {
    hooks.release();
    hooks.test();
    hooks.build();
  },
  task: function() {
    grunt.config.merge(helpers.getTaskConfig([
      'shell:deleteCoverages',
      'shell:bower',
      'jshint',
      'ngtemplates:dist',
      'karma:all',
      'shell:updateWebdriver',
      'shell:deleteInstrumented',
      'instrument',
      'makeReport',
      'connect:test',
      'protractor_coverage:single',
      'update-contributors',
      'bump',
      'less:dist',
      'sass:dist',
      'concat:dist',
      'concat:distStyles',
      'ngAnnotate:dist',
      'uglify:dist',
      'concat:distStylesMin',
      'cssmin:dist'
    ]));
    grunt.registerTask('release', availableTasks.release.desc, function(type) {
      grunt.task.run([
        'release:before',
        'test',
        'update-contributors',
        'bump-only:' + (type || 'patch'),
        'build',
        'bump-commit',
        'release:after'
      ]);
    });
  }
};


module.exports = {
  addTasks: function(tasksToAdd) {
    if (typeof tasksToAdd === 'undefined') {
      tasksToAdd = Object.keys(availableTasks);
    }

    grunt.registerTask(
      '_normalizeUnitCoverage',
      'helper for coverage reports *',
      helpers.normalizeCoverageTask
    );

    tasksToAdd.forEach(function(task) {
      if (availableTasks[task]) {
        availableTasks[task].hooks();
      } else {
        grunt.log.error('undefined task "' + task + '" in grunt-angular-toolbox');
      }
    });

    tasksToAdd.forEach(function(task) {
      if (availableTasks[task]) {
        availableTasks[task].task();
      }
    });
  },
  helpers: helpers
};
